### PR TITLE
PLATFORM-5998: adding apple provider

### PR DESCRIPTION
**Pull request by themech**
Adding dedicated Apple provider. Unfortunately Apple uses JWT tokens as secrets so the generic provider does not work.
Also adding the possibility for the callback to be called using POST.

### CODE REVIEW

File: driver/config/.schema/config.schema.json [hunk 1]
```diff
@@ -355,7 +380,6 @@
         "id",
         "provider",
         "client_id",
-        "client_secret",
```
Comment (themech) [inline driver/config/.schema/config.schema.json:358]: client secret is not required for apple. So I'm adding it as a required field in a condition below

File: selfservice/strategy/oidc/provider.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ type Claims struct {
 	Picture             string `json:"picture,omitempty"`
 	Website             string `json:"website,omitempty"`
 	Email               string `json:"email,omitempty"`
-	EmailVerified       bool   `json:"email_verified,omitempty"`
+	EmailVerified       bool   `json:"email_verified,string,omitempty"`
```
[the remark this answers is not in the document]
Comment (harnash) [reply selfservice/strategy/oidc/provider.go:30]: won't this break other providers that properly encode this as a boolean?
https://play.golang.org/p/eezg4cj7YD0

[the remark this answers is not in the document]
Comment (themech) [reply selfservice/strategy/oidc/provider.go:30]: Right. It looks like we'll need a custom `UnmarshalJSON` then :(

[the remark this answers is not in the document]
Comment (themech) [reply selfservice/strategy/oidc/provider.go]: Added a new `ConvertibleBoolean` type to handle both booleans and strings when unmarshalling. AFAIK the json first go to kratos and only then to jsonnet so we have to handle it on the golang site

File: selfservice/strategy/oidc/provider.go [hunk 3]
```diff
@@ -13,28 +15,44 @@ type Provider interface {
 	AuthCodeURLOptions(r ider) []oauth2.AuthCodeOption
 }
 
+// ConvertibleBoolean can unmarshal both booleans and strings.
+// This is needed as Apple casually sends the email_verified field as a string.
+type ConvertibleBoolean bool
```
Comment (themech) [inline selfservice/strategy/oidc/provider.go:20]: not sure if to add this type here next to where it's used or in a separate file

Comment (harnash) [reply selfservice/strategy/oidc/provider.go:20]: Looks like general purpose type. I'm slightly leaning towards putting it into some commonly accessible package.

Comment (themech) [reply selfservice/strategy/oidc/provider.go:20]: Like put it in https://github.com/Wikia/kratos/tree/master/x?

Comment (harnash) [reply selfservice/strategy/oidc/provider.go:20]: Yeah. I guess.

File: selfservice/strategy/oidc/provider.go [hunk 4]
```diff
@@ -13,28 +15,44 @@ type Provider interface {
...
+
+func (bit *ConvertibleBoolean) UnmarshalJSON(data []byte) error {
+	asString := string(data)
+	if asString == "true" || asString == "\"true\"" {
```
Comment (harnash) [inline selfservice/strategy/oidc/provider.go:24]:
```suggestion
	if asString == "true" || asString == `"true"` {
```

File: selfservice/strategy/oidc/provider.go [hunk 5]
```diff
@@ -13,28 +15,44 @@ type Provider interface {
...
+		*bit = true
+	} else if asString == "false" || asString == "\"false\"" {
```
Comment (harnash) [inline selfservice/strategy/oidc/provider.go:26]:
```suggestion
	} else if asString == "false" || asString == `"false"` {
```

File: selfservice/strategy/oidc/provider_apple.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+func (a *ProviderApple) newClientSecret() (string, error) {
+	// decode the pem format
+	block, _ := pem.Decode([]byte(a.config.PrivateKey))
+	if block == nil || block.Type != "PRIVATE KEY" {
+		return "", errors.New("failed to decode PEM block containing private key")
+	}
+
+	parsedKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
+		return "", errors.New("Private key decoding failed. " + err.Error())
```
Comment (harnash) [inline selfservice/strategy/oidc/provider_apple.go:43]:
```suggestion
		return "", errors.Wrap(err.Error(), "Private key decoding failed")
```

**Review by harnash — CHANGES_REQUESTED** [review]

**Review by emkarcinos — APPROVED** [review]

**Review by harnash — APPROVED** [review]

File: selfservice/strategy/oidc/provider_apple.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+func (a *ProviderApple) OAuth2(context.Context) (*oauth2.Config, error) {
+	return a.oauth2()
+}
+
+func (a *ProviderApple) AuthCodeURLOptions(r ider) []oauth2.AuthCodeOption {
+	var options []oauth2.AuthCodeOption
+
+	if isForced(r) {
+		options = append(options, oauth2.SetAuthURLParam("prompt", "login"))
```
Comment (abador) [inline selfservice/strategy/oidc/provider_apple.go:90]: Is this really needed?

Comment (abador) [reply selfservice/strategy/oidc/provider_apple.go:90]: It seems that changing this doesn't do anything and apple forces authentication. The flow seems to be ok from the technical perspective. As a user I would wonder what is happening and why do I need to reauthenticate

**Pull request merged**